src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIMarsHandle.cs

```diff
@@ -449,15 +468,19 @@ public void HandleAck(uint highwater)
         /// <param name="header">SMUX header</param>
         public void HandleReceiveComplete(SNIPacket packet, SNISMUXHeader header)
         {
+            if (Monitor.IsEntered(_connection.DemuxerSync))
+            {
+                throw new InvalidOperationException("SNIMarsHandle.HandleRecieveComplete should be be called while holding the SNIMarsConnection.DemuxerSync because it can cause deadlocks");
+            }
             long scopeID = SqlClientEventSource.Log.TrySNIScopeEnterEvent(s_className);
             try
             {
                 lock (this)
```

Comment:
not related to your changes, but in general `lock(this)` is not a proper way of locking. When we use `This` unrelated areas of the code/driver are using the same object for locking. Should not we have a lock object in this situation or similar situations?

Reply:
I investigated this. The reason that `lock (this)` is advised against is that if the object is public then unexpected users can use the same object to lock on causing unexpected behaviour. The suggested alternative is using dedicated and private lock object. The same reason applies for not locking on types. In the situation we have here we know that the object is never exposed to users outside our library and we know that we are being careful about locking and lifetimes. So the guidance is good but we don't need to follow it because we have control.
However, adding an additional dedicated locking object isn't expensive if you and others prefer, i just thought it wasn't necessary.

Reply:
I would vote for a dedicated private lock for two reasons:
1. We follow the guideline, although there is no harm done by its current implementation.
2.  We don't know what changes will be forced, by issues fixation or new features coming in.... I am waiting for `Windows 11` to kick in. Things may move around and we may have to give access to some of our APIs, most probably that will never  happen in this case, but we make our own pattern and style, to use a a dedicated private lock everywhere inside the driver.

Reply:
Windows 11 isn't going to change anything at this level. It's an OS update and if it changes dotnet runtime behaviours it'd be a massive ecosystem breaking change. That said, I'm ok with adding a dedicated sync object so that's done.